File: conways_rust/src/conways.rs
// -----------------------------------------------------------------------------
// conways.rc : Conways game of life in Rust
// -----------------------------------------------------------------------------

extern crate rand;
extern crate array2d;

use rand::Rng;
use std::env;
use array2d::Array2D;
use std::fs::File;
use std::io::{BufWriter, Write};
use std::time::*;

fn write_file(name:&String, rows:usize, cols:usize, world: &mut Array2D<i32>) {
    let f = File::create(name).expect("Unable to create file");
    let mut f = BufWriter::new(f);
    for i in 0..rows {
        for j in 0..cols {
            let to_write = format!("{} ",  world[(i,j)]);
            f.write(to_write.as_bytes()).expect("Unable to write data");
        }
        let to_write = format!("{}",  "\n");
        f.write_all(to_write.as_bytes()).expect("Unable to write data");
    }
}

fn init_world(rows:usize,cols:usize,world: &mut Array2D<i32>) {
    let mut rng = rand::thread_rng();
    for i in 0..rows {
        for j in 0..cols {
            world[(i, j)] = rng.gen_range(0..2);
        }
    }
}

fn update_world(rows:usize,cols:usize,world: &mut Array2D<i32>) {    
    
    let old_world = world.clone();
    
    for i in  0..rows {
        for j in  0..cols {
            let state = old_world[(i, j)];                 
            let neis = old_world[((i).rem_euclid(rows), (j + 1).rem_euclid(cols))]
                      +old_world[((i).rem_euclid(rows), (j - 1).rem_euclid(cols))]
                      +old_world[((i + 1).rem_euclid(rows), (j).rem_euclid(cols))]
                      +old_world[((i - 1).rem_euclid(rows), (j).rem_euclid(cols))]
                      +old_world[((i + 1).rem_euclid(rows), (j + 1).rem_euclid(cols))]
                      +old_world[((i - 1).rem_euclid(rows), (j - 1).rem_euclid(cols))]
                      +old_world[((i + 1).rem_euclid(rows), (j - 1).rem_euclid(cols))]
                      +old_world[((i - 1).rem_euclid(rows), (j + 1).rem_euclid(cols))];
            
            if state == 1 {
                if neis != 2 && neis != 3 {
                    world[(i, j)] = 0;
                }
            }
            else {
                if neis == 3 {
                    world[(i, j)] = 1;
                }  
            } 
        }
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let rows: usize = args[1].parse().unwrap();
    let cols: usize = args[2].parse().unwrap();
    let frames: usize = args[3].parse().unwrap();    
    
    let world = &mut Array2D::filled_with(0, rows, cols);
    init_world(rows, cols, world);

    let name = format!("{:06}.txt", 0);
    write_file(&name, rows, cols, world);

    println!("Started simulation");
    let now = Instant::now();
    for frame in 1..frames + 1
    {
        update_world(rows, cols, world);
        let name = format!("{:06}.txt", frame);
        write_file(&name, rows, cols, world);
    }
    let elapsed_time = now.elapsed();
    println!("Finished simulation in: {:6.2} seconds", 
    elapsed_time.as_millis() as f32/1000.0);
}

File: conways_go/conways.go
// -----------------------------------------------------------------------------
// conways.go : Conways game of life in Go (5 min 11 sec)
// -----------------------------------------------------------------------------

package main

import (
    "fmt"
    "os"
    "strconv"
    "math/rand"
    "time"
)

func mod(a, b int) int {
    return ((a % b) + b) % b
}

func write_file(name string, rows int, cols int, world []int) {

    f, _ := os.Create(name)
    for i := 0; i < rows; i++ {
        for j := 0; j < cols; j++ {
            fmt.Fprintf(f, "%d ", world[i * cols + j])
        }
        fmt.Fprintf(f, "\n")
    }
}

func init_world(rows int, cols int) []int {

    s1 := rand.NewSource(time.Now().UnixNano())
    r1 := rand.New(s1)
    world := make([]int, rows*cols)

    for i := 0; i < rows; i++ {
        for j := 0; j < cols; j++ {
            world[i * cols + j] = r1.Intn(2)
        }
    }
    return world
}

func update_world(rows int, cols int, world []int) {

    old_world := make([]int, rows*cols)
    _ = copy(old_world, world)

    for i := 0; i < rows; i++ {
        for j := 0; j < cols; j++ {
            state := old_world[i * cols + j];                 
            neis := old_world[mod(i, rows) * cols + mod(j - 1, rows)] + 
                    old_world[mod(i, rows) * cols + mod(j + 1, rows)] + 
                    old_world[mod(i + 1, rows) * cols + mod(j, rows)] + 
                    old_world[mod(i - 1, rows) * cols + mod(j, rows)] +
                    old_world[mod(i + 1, rows) * cols + mod(j + 1, rows)] +
                    old_world[mod(i - 1, rows) * cols + mod(j - 1, rows)] +
                    old_world[mod(i + 1, rows) * cols + mod(j - 1, rows)] +
                    old_world[mod(i - 1, rows) * cols + mod(j + 1, rows)]

            if state == 1 {
                if neis != 2 && neis != 3 {
                    world[i * cols + j] = 0
                }
            } else {
                if neis == 3 {
                    world[i * cols + j] = 1
                }  
            }
        }
    }
}

func main() {

    rows, _ := strconv.Atoi(os.Args[1])
    cols, _ := strconv.Atoi(os.Args[2])
    frames, _ := strconv.Atoi(os.Args[3])
    world := init_world(rows, cols)

    name := fmt.Sprintf("%06d.txt", 0)
    write_file(name, rows, cols, world)

    fmt.Printf("Started simulation\n")
    start := time.Now()
    for frame := 1; frame < frames + 1; frame++ {
        update_world(rows, cols, world)
        name = fmt.Sprintf("%06d.txt", frame)
        write_file(name, rows, cols, world)
    }
    duration := time.Since(start)
    fmt.Printf("Finished simulation in: %6.2f seconds\n", duration.Seconds())
}